test/job-test.js

```diff
@@ -0,0 +1,299 @@
+
+            try {
+                bigQueryJob.getQuerySQL();
+                expect.fail('getQuerySQL() should have failed');
+            }
+            catch(err) {
+                expect(err.code).to.equals('ERROR_TEMPLATE');
+                expect(err.message).to.equals('dataset is not defined');
+                expect(err.stack).to.contains('TemplateException: ReferenceError: dataset is not defined');
+            }
+        });
+
+        it('2. Build query with only [dataset] variable. Expect to throw exception ', async () => {
+
+            try {
+                const getQueryParamsStub = sinon.stub(bigQueryJob, 'getQueryParams').returns({ dataset: 'DATASET' });
+                bigQueryJob.getQuerySQL();
+                expect.fail('getQuerySQL() should have failed');
+            }
+            catch(err) {
+                expect(err.code).to.equals('ERROR_TEMPLATE');
+                expect(err.message).to.equals('table is not defined');
+                expect(err.stack).to.contains('TemplateException: ReferenceError: table is not defined');
+            }
+            finally {
+                sinon.restore();
+            }
+        });
+
+        it('3. Build query with variables. Expect to build query successfully ', async () => {
+
+            const getQueryParamsStub = sinon.stub(bigQueryJob, 'getQueryParams').returns({ dataset: 'DATASET', table: 'TABLE' });
+            const querySQL = bigQueryJob.getQuerySQL();
+
+            expect(querySQL).to.equals('SELECT some_field, other_field FROM `DATASET.TABLE`');
+            getQueryParamsStub.restore();
+        });
+
+        it('4. Build query with variables and debug string. Expect to build query successfully ', async () => {
+
+            const getQuerySQLDebugStub = sinon.stub(bigQueryJob, 'getQuerySQLDebug').returns('-- JobId: JOB_ID\n-- JobName: JOB_NAME');
+            const getQueryParamsStub = sinon.stub(bigQueryJob, 'getQueryParams').returns({ dataset: 'DATASET', table: 'TABLE' });
+            const querySQL = bigQueryJob.getQuerySQL();
+
+            expect(querySQL).to.equals('-- JobId: JOB_ID\n-- JobName: JOB_NAME\nSELECT some_field, other_field FROM `DATASET.TABLE`');
+            getQueryParamsStub.restore();
+        });
+    });
+
+    describe('3. getQueryOptions()', () => {
+
+        let bigQueryJob = null;
+
+        before(async () => {
+            const opts = _getOptions();
+            bigQueryJob = new BigQueryJob(opts);
+            await bigQueryJob.init();
+        });
+
+        beforeEach(() => {
+            sinon.restore();
+        });
+
+        after(() => {
+            sinon.restore();
+        });
+
+        it('1. Get query params with default values. Expect to return correct values with defaults', () => {
+            const getQueryParamsStub = sinon.stub(bigQueryJob, 'getQueryParams').returns({ dataset: 'DATASET', table: 'TABLE' });
```

Comment:
Same here, a subclass of BigQueryJob will make the following examples more clear to me.